File: 2025/src/main/java/dpr/aoc2025/Day01.java
package dpr.aoc2025;

import java.util.List;

import dpr.commons.Day;
import dpr.commons.Util;

class Day01 implements Day {
    @Override
    public void execute() {
        var lines = Util.getNotEmptyLinesFromFile(dayNum(), "input.txt");
        System.out.println(part1(lines));
        System.out.println(part2(lines));
    }

    @Override
    public int dayNum() {
        return 1;
    }

    int part1(List<String> lines) {
        return calculateZeroOccurrences(lines, false);
    }

    int part2(List<String> lines) {
        return calculateZeroOccurrences(lines, true);
    }

    private static int calculateZeroOccurrences(List<String> lines, boolean countPassing) {
        int pos = 50;
        int count = 0;
        int limit = 100;
        for (String line : lines) {
            var dir = line.charAt(0);
            var move = Integer.parseInt(line.substring(1));
            if (countPassing) {
                count += move / limit;
            }
            move %= limit;
            if (dir == 'L') {
                int intermediate = pos - move;
                if (countPassing && pos != 0 && intermediate < 0) {
                    ++count;
                }
                pos = (pos - move + limit) % limit;
            } else {
                int intermediate = pos + move;
                if (countPassing && intermediate > limit) {
                    ++count;
                }
                pos = (pos + move + limit) % limit;
            }
            if (pos == 0) {
                ++count;
            }
        }
        return count;
    }

    public static void main(String[] args) {
        new Day01().execute();
    }
}

File: 2021/src/main/kotlin/dpr/aoc2021/Day01.kt
package dpr.aoc2021

import dpr.commons.Util

object Day01 {
    @JvmStatic
    fun main(args: Array<String>) = Util.measureTime {
        val input = Util.getNotEmptyLinesFromFile("/01/input.txt")
        val numbers = parseInput(input)

        println(part1(numbers))
        println(part2(numbers))
    }

    @JvmStatic
    fun parseInput(input: List<String>): List<Int> = input.map { it.toInt() }

    @JvmStatic
    fun part1(numbers: List<Int>): Int {
        return numbers.zipWithNext().count { it.first < it.second }
    }

    @JvmStatic
    fun part2(numbers: List<Int>): Int {
        return numbers.windowed(3, 1).map { it.sum() }.zipWithNext().count { it.first < it.second }
    }
}

